crypto.cpp

```diff
@@ -123,126 +133,484 @@ OvpnCryptoUninitAlgHandles(_In_ BCRYPT_ALG_HANDLE aesAlgHandle, BCRYPT_ALG_HANDL
-
      [DATA_V2 opcode] [Packet ID] [AEAD Auth tag] [ciphertext]
-     [4 bytes       ] [4/8 bytes] [16 bytes     ]
+     [4 bytes       ] [4 bytes  ] [16 bytes     ]
      [AEAD additional data(AD)  ]
 
-    With CRYPTO_OPTIONS_AEAD_TAG_END AEAD Auth tag is placed after ciphertext:
+    New data format, with epoch keys and 64bit packet id:
+
+     struct aead_packet {
+        int opcode:5;
+        int key_id:3;
+        int peer_id:24;
+        uint64_t packet_id;
+        uint8_t* encrypted_payload;
+        uint8_t[16] authentication_tag;
+     }
+
+     struct packet_id {
+        uint epoch:16;
+        uint epoch_counter:48;
+     }
+
+     authenticated_data = opcode| key_id | peer_id | packet_id
 
-     [DATA_V2 opcode] [Packet ID] [ciphertext] [AEAD Auth tag]
-     [4 bytes       ] [4/8 bytes]              [16 bytes     ]
-     [AEAD additional data(AD)  ]
     */
 
     NTSTATUS status = STATUS_SUCCESS;
 
-    BOOLEAN pktId64bit = cryptoOptions & CRYPTO_OPTIONS_64BIT_PKTID;
-
-    SIZE_T cryptoOverhead = OVPN_DATA_V2_LEN + AEAD_AUTH_TAG_LEN + (pktId64bit ? 8 : 4);
+    BOOLEAN authTagEnd = opts->UseEpoch;
+    ULONG pktidLen = opts->UseEpoch ? 8 : 4;
+    ULONG cryptoOverhead = OVPN_DATA_V2_LEN + AEAD_AUTH_TAG_LEN + pktidLen;
 
     if (len < cryptoOverhead) {
         LOG_WARN("Packet too short", TraceLoggingValue(len, "len"));
         return STATUS_DATA_ERROR;
     }
 
+    // we prepended buf with crypto overhead
+    len -= cryptoOverhead;
+
+    OvpnCryptoKeyContext* decryptKey = &keySlot->Decrypt;
+    UINT64 packet_id = 0;
+    UINT16 rx_epoch = 0;
+
     UCHAR nonce[12];
     if (encrypt) {
         // prepend with opcode, key-id and peer-id
         UINT32 op = OvpnProtoOp32Compose(OVPN_OP_DATA_V2, keySlot->KeyId, keySlot->PeerId);
         op = RtlUlongByteSwap(op);
-        *reinterpret_cast<UINT32*>(bufOut) = op;
+        RtlCopyMemory(bufOut, &op, sizeof(op));
 
-        if (pktId64bit)
-        {
-            // calculate pktid
-            UINT64 pktid;
-            GOTO_IF_NOT_NT_SUCCESS(done, status, OvpnPktidXmitNext(&keySlot->PktidXmit, &pktid, true));
-            ULONG64 pktidNetwork = RtlUlonglongByteSwap(pktid);
+        if (opts->UseEpoch) {
+            if (keySlot->EpochKeySend.Epoch == UINT16_MAX) {
+                return STATUS_BUFFER_OVERFLOW;
+            }
 
-            // calculate nonce, which is pktid + nonce_tail
-            RtlCopyMemory(nonce, &pktidNetwork, 8);
-            RtlCopyMemory(nonce + 8, keySlot->EncNonceTail, 4);
+            if (OvpnCryptoAeadUsageLimitReached(opts->AeadUsageLimit, keySlot->Encrypt.PlaintextBlocks, keySlot->PktidXmit.SeqNum) || (keySlot->PktidXmit.SeqNum == PACKET_ID_EPOCH_MAX)) {
+                OvpnCryptoEpochIterateSendKey(keySlot, opts);
+            }
+
+            // calculate 64-bit packet-id = (epoch << 48) | ctr48
+            UINT64 ctr48 = InterlockedIncrementNoFence64(&keySlot->PktidXmit.SeqNum) & 0x0000FFFFFFFFFFFFULL;
```

Comment:
There is no overflow checky here, you are just masking the bits. This need to either checked here or a comment to say where we ensure that this never overflows.